logstash-rs/src/buffer.rs

```diff
@@ -0,0 +1,174 @@
+        sender: S,
+        buffer_size: Option<usize>,
+        buffer_lifetime: Option<Duration>,
+        ignore_buffer: Level,
+    ) -> Self {
+        let buffer_size = match buffer_size {
+            Some(s) if s < 2 => None,
+            x => x,
+        };
+        Self {
+            sender,
+            buffer: Vec::with_capacity(buffer_size.unwrap_or(0)),
+            buffer_size,
+            buffer_lifetime,
+            deadline: None,
+            ignore_buffer,
+        }
+    }
+
+    fn run(self) -> mpsc::SyncSender<Command> {
+        let (sender, receiver) = mpsc::sync_channel(1);
+        self.run_thread(receiver);
+        sender
+    }
+
+    fn next_deadline(&self) -> Option<Instant> {
+        if self.buffer.is_empty() && self.buffer_size.is_some() {
+            return self.buffer_lifetime.map(|lt| Instant::now() + lt);
+        }
+        None
+    }
+
+    fn run_thread(mut self, receiver: mpsc::Receiver<Command>) {
+        std::thread::spawn::<_, Result<()>>(move || {
+            {
+                loop {
+                    let cmd = match self.deadline {
+                        Some(deadline) => receiver
+                            .recv_timeout(deadline.saturating_duration_since(Instant::now())),
+                        None => receiver
+                            .recv()
+                            .map_err(|_| mpsc::RecvTimeoutError::Disconnected),
+                    };
+
+                    if let Ok(Command::SendBatch(_) | Command::Send(_)) = &cmd {
+                        self.deadline = self.next_deadline();
+                    }
+                    match cmd {
+                        Ok(Command::Flush) | Err(mpsc::RecvTimeoutError::Timeout) => {
+                            self.flush()?
+                        }
+                        Ok(Command::Send(event)) => self.send(event)?,
+                        Ok(Command::SendBatch(events)) => self.send_batch(events)?,
+                        Err(mpsc::RecvTimeoutError::Disconnected) => break,
+                    }
+                }
+                Ok(())
+            }
+            .map_err(|err| {
+                println!("fatal logger error: {}", err);
+                err
+            })
+        });
+    }
+
+    fn send(&mut self, event: LogStashRecord) -> Result<()> {
+        if event.level >= self.ignore_buffer {
+            self.sender.send(event)?;
+        } else if let Some(max_size) = self.buffer_size {
+            self.buffer.push(event);
+            if self.buffer.len() >= max_size {
+                self.flush()?;
+            }
+        } else {
+            self.sender.send(event)?;
+        }
+        Ok(())
+    }
+
+    fn send_batch(&mut self, events: Vec<LogStashRecord>) -> Result<()> {
+        for event in events {
+            self.send(event)?;
+        }
+        Ok(())
+    }
+
+    fn flush(&mut self) -> Result<()> {
+        if !self.buffer.is_empty() {
+            let buffer = std::mem::replace(
```

Comment:
You access `self.buffer` from different threads, but `std::mem::replace` is not a thread-safe method. You need to protect `buffer` with synchronization primitive

Reply:
@idruzhitskiy explained me, that everything is ok here